libtock/alarm_timer.c

```diff
@@ -4,11 +4,131 @@
 #include <limits.h>
 #include <stdlib.h>
 
+#define MAX_TICKS UINT32_MAX
+
 // Returns 0 if a <= b < c, 1 otherwise
 static int within_range(uint32_t a, uint32_t b, uint32_t c) {
   return (b - a) < (b - c);
 }
 
+static uint32_t ticks_to_ms(uint32_t ticks) {
+  // `ticks_to_ms`'s conversion will be accurate to within the range
+  // 0 to 1 milliseconds less than the exact conversion
+  // (true millisecond conversion - [0,1) milliseconds).
+
+  uint32_t frequency;
+  alarm_internal_frequency(&frequency);
+
+  uint32_t seconds = (ticks / frequency);
+  uint32_t milliseconds_per_second = 1000;
+
+  // Calculate the conversion of full seconds to ticks.
+  uint32_t milliseconds = seconds * milliseconds_per_second;
+
+  // To get conversion accuracy within 1 millisecond, the conversion
+  // must also convert partial seconds.
+  uint32_t leftover_ticks = ticks % frequency;
+
+  // This calculation is mathematically equivalent to doing:
+  //
+  //   `leftover_ticks` / (`frequency` / `milliseconds_per_second`)
+  //
+  // This is taking the remaining unconverted ticks (leftover_ticks)
+  // and dividing by the number of ticks per millisecond
+  // (`frequency` (ticks per second) / `1000` milliseconds per second)
+  // The division is done this way because of the same argument in
+  // `ms_to_ticks`.
+  milliseconds += (leftover_ticks * milliseconds_per_second) / frequency;
```

Comment:
`leftover_ticks` is at most equal to `frequency`. For frequency values larger than 4MHz, I believe this calculation will overflow. I do not think there is a way around having a `uint64_t` to handle the intermediate calculations.

Using a `uint64_t` would allow us to support up to and beyond GHz frequencies.

Reply:
Making the calculation use `uint64_t`s will not solve this issue because the timer interface is 32-bit. This function *must* return a 32-bit integer. It is a separate concern whether a 64-bit timer interface should be added.

Reply:
I'm just browsing through all this, and I think I agree with the logic here. Using `uint64_t` would also allow us to simply the logic here a bunch, and remove the split between seconds and fractional seconds, effectively making this equivalent to: `ticks_to_ms = (ticks * 1000) / freq`.

I'm wondering how much the split into seconds and fractional seconds buys us here in terms of performance in code size, vs. 64-bit arithmetic on a 32-bit platform.

Reply:
Nevermind, I just saw that the new, merged PR has been changed to 64-bit, my bad. Part of my comment still applies, I suppose: https://github.com/tock/libtock-c/pull/434/files#r1642913413